python/lsst/ip/diffim/zogy.py

```diff
@@ -898,3 +949,133 @@ class ZogyMapReduceConfig(ImageMapReduceConfig):
+        config.cellSizeX = config.cellSizeY = 20
+        config.borderSizeX = config.borderSizeY = 6
+        config.reducerSubtask.reduceOperation = 'average'
+
+    def _computeImageMean(self, exposure):
+        """Compute the sigma-clipped mean of the pixels image of `exposure`.
+        """
+        statsControl = afwMath.StatisticsControl()
+        statsControl.setNumSigmaClip(3.)
+        statsControl.setNumIter(3)
+        ignoreMaskPlanes = ("INTRP", "EDGE", "DETECTED", "SAT", "CR", "BAD", "NO_DATA", "DETECTED_NEGATIVE")
+        statsControl.setAndMask(afwImage.MaskU.getPlaneBitMask(ignoreMaskPlanes))
+        statObj = afwMath.makeStatistics(exposure.getMaskedImage().getImage(),
+                                         exposure.getMaskedImage().getMask(),
+                                         afwMath.MEANCLIP|afwMath.MEDIAN, statsControl)
+        mn = statObj.getValue(afwMath.MEANCLIP)
+        med = statObj.getValue(afwMath.MEDIAN)
+        return mn, med
+
+    def subtractExposures(self, templateExposure, scienceExposure,
+                          doWarping=True, spatiallyVarying=True, inImageSpace=False,
+                          doPreConvolve=False):
+
+        mn1 = self._computeImageMean(templateExposure)
+        mn2 = self._computeImageMean(scienceExposure)
+        self.log.info("Exposure means=%f, %f; median=%f, %f:" % (mn1[0], mn2[0], mn1[1], mn2[1]))
+        if not np.isnan(mn1[0]) and np.abs(mn1[0]) > 1:
+            mi = templateExposure.getMaskedImage()
+            mi -= mn1[0]
+        if not np.isnan(mn2[0]) and np.abs(mn2[0]) > 1:
+            mi = scienceExposure.getMaskedImage()
+            mi -= mn2[0]
+
+        self.log.info('Running Zogy algorithm: spatiallyVarying=%r' % spatiallyVarying)
+
+        if not self._validateWcs(templateExposure, scienceExposure):
+            if doWarping:
+                self.log.info("Astrometrically registering template to science image")
+                # Also warp the PSF
+                xyTransform = afwImage.XYTransformFromWcsPair(scienceExposure.getWcs(),
+                                                              templateExposure.getWcs())
+                psfWarped = measAlg.WarpedPsf(templateExposure.getPsf(), xyTransform)
+                templateExposure = self._warper.warpExposure(scienceExposure.getWcs(),
+                                                             templateExposure,
+                                                             destBBox=scienceExposure.getBBox())
+
+                templateExposure.setPsf(psfWarped)
+                templateExposure.writeFits('WARPEDTEMPLATE_ZOGY.fits')
+            else:
+                self.log.error("ERROR: Input images not registered")
+                raise RuntimeError("Input images not registered")
+
+        def gm(exp):
+            return exp.getMaskedImage().getMask()
+        def ga(exp):
+            return exp.getMaskedImage().getImage().getArray()
+        def gv(exp):
```

Comment:
Also, user convenience functions like these make the code harder to read.  Please don't do this.

Also, with pybind11, you can write this as e.g. `exp.maskedImage.mask` which might make this more palatable.